test/lib.sh

```diff
@@ -59,3 +59,98 @@ EOF
 hr() {
     echo "========================="
 }
+
+## diff BASELINE
+#
+# Recursively compare the current repository content to the BASELINE repository
+#
+# :param BASELINE: An existing directory used as reference to compared sync'ed files
+diff() {
+	for file in `ls` ; do 
+		if [ -d $file ] ; then 
+			pushd $file  > /dev/null
+			diff $1/$file 
+			popd  > /dev/null
+		elif [ -f $file ] ; then
+			cmp $file $1/$file
+			if [ ! $? = 0 ] ; then
+				handle_error_counter INC
+			fi
+		fi
+	done
+}
+
+## compare FOLDER
+#
+# Check FOLDER is properly sync'ed, determining the reference base on FOLDER 
+#
+# :param FOLDER: An existing directory sync'ed by boilerplate and to be checked
+compare() {
+	if [ $1 = "_data" ] ; then
+		if [ ! -f $repo/boilerplate/_data/last_boilerplate_commit ] ; then
+			# TODO: Check the content of the file to ensure it contains the proper commit in addition to the file existence
+			handle_error_counter INC
+		fi
+	else
+		pushd $1  > /dev/null
+		diff $BOILERPLATE_GIT_REPO/boilerplate/$1
+		popd > /dev/null
+	fi
+}
+
+## check_update
+#
+# Check the boilerplate synchronization is properly working, covering generics and convention
+# specific parts
+check_update() {
+	handle_error_counter INIT
+	pushd $repo/boilerplate > /dev/null
+	
+	compare _data
+	compare _lib
+	
+	while read convention ; do
+	  if [ -d $BOILERPLATE_GIT_REPO/boilerplate/$convention ] ; then
+		  compare $convention
+	  fi
+	done < $repo/boilerplate/update.cfg
+	
+	popd > /dev/null
+	
+	handle_error_counter CHECK
+	
+	exit $?
+}
+
+## add_convention CONVENTION
+#
+# Add a convention and run the update script for the project to pick it up
+# :param CONVENTION: An existing convention
+add_convention() {
+	if grep -q "^$1\$" $repo/boilerplate/update.cfg ; then
+		echo $1 >> $repo/boilerplate/update.cfg
+	fi
+}
+
+## handle_error_counter FUNCTION
+#
+# Function managing the error counter to increment in case of diff found for a file and print before returning
+# :param FUNCTION: action to be done on the error counter
+#  - RESET : Initilize the error counter to 0 
+#  - INC : Increment the internal counter by 1
+#  - CHECK : Check the number of diff is 0. If it is not, returns 1 and print the number of errors found
+handle_error_counter() {
+	declare -i error_counter
```

Comment:
This is declaring a local variable. It goes out of scope as soon as the function returns. So after fixing the `exit`s below:

```
[efried@efried boilerplate]$ handle_error_counter RESET
[efried@efried boilerplate]$ handle_error_counter INC
[efried@efried boilerplate]$ handle_error_counter CHECK
[efried@efried boilerplate]$ echo $?
0
```